collections: Add function to partition page slices

raoulb opened the pull request:
Another thing I think is missing from the standard palette of template functions: partition a slice (of pages) into a set of equal chunks (plus a possible remainder).

This implements the proposal from issue [#11553](https://github.com/gohugoio/hugo/issues/11553).

Help wanted to discuss the implementation details. I'll also add docs and tests.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/gohugoio/hugo?pullRequest=12661) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/gohugoio/hugo?pullRequest=12661) before we can accept your contribution.<br/><hr/>**Ral** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/gohugoio/hugo?pullRequest=12661) it.</sub>

bep commented:
Thanks for this. Issue #11553 is about a bug in `slice`. This PR does not fix that particular bug. I don't say this isn't a great idea, but we cannot start the discussion with a implemention. Please opne a new proposal if you want to discuss this.

raoulb commented:
This PR does not directly address the change in behavior of `append` combined with `slice` as shown in bug #11131 (which broke several sites). For that, one should first decide if this change is actually a bug or just the new correct behavior. (Last time I've read that report it appeared to me tendency was towards the latter.)

However, the proposed new `partition` functionality here directly addresses my own bug report #11553 which is very likely related to the first one.

I just updated the branch to include some unit tests. This code is now no longer WiP from my point of view and I'd like to know what needs to be changed to get this merged. (Should I open a fresh PR?)

In any case, I propose to add to Hugo the partition functionality with the semantics as documented and implemented by this code.
